File: design_patterns/04_Factory/data_reader.js
const fs = require('fs');
const path = require('path');

// Abstract class for DataReader
class DataReader {
    readData(file_path) {}
}

// Concrete implementations of DataReader
class CSVDataReader extends DataReader {
    readData(file_path) {
        const data = fs.readFileSync(file_path, 'utf8').split('\n').map(row => row.split(','));
        return data;
    }
}

class JSONDataReader extends DataReader {
    readData(file_path) {
        const data = JSON.parse(fs.readFileSync(file_path, 'utf8'));
        return data;
    }
}

class XMLDataReader extends DataReader {
    readData(file_path) {
        const xmlContent = fs.readFileSync(file_path, 'utf8');
        const data = this.parseXML(xmlContent);
        return data;
    }

    parseXML(xmlContent) {
        const parser = new Parser();
        const data = {};
        parser.parseString(xmlContent, (err, result) => {
            if (err) {
                console.error(err);
                return;
            }
            data.root = result;
        });
        return data;
    }
}

// Factory class for creating data readers
class DataReaderFactory {
    static createDataReader(file_path) {
        const file_extension = path.extname(file_path).toLowerCase();
        if (file_extension === ".csv") {
            return new CSVDataReader();
        } else if (file_extension === ".json") {
            return new JSONDataReader();
        } else if (file_extension === ".xml") {
            return new XMLDataReader();
        } else {
            throw new Error("Unsupported file type");
        }
    }
}

// Client code
if (process.argv.length !== 3) {
    console.log("Usage: node script.js <file_path>");
    process.exit(1);
}

const file_path = process.argv[2];

const dataReader = DataReaderFactory.createDataReader(file_path);
const data = dataReader.readData(file_path);
console.log("Data read from", file_path, ":", data);

File: design_patterns/04_Factory/data_reader.py
import csv
import json
import xml.etree.ElementTree as ET
from abc import ABC, abstractmethod
import os
import sys

# Abstract class for DataReader
class DataReader(ABC):
    @abstractmethod
    def read_data(self, file_path):
        pass

# Concrete implementations of DataReader
class CSVDataReader(DataReader):
    def read_data(self, file_path):
        data = []
        with open(file_path, 'r') as file:
            csv_reader = csv.reader(file)
            for row in csv_reader:
                data.append(row)
        return data

class JSONDataReader(DataReader):
    def read_data(self, file_path):
        with open(file_path, 'r') as file:
            data = json.load(file)
        return data

class XMLDataReader(DataReader):
    def read_data(self, file_path):
        tree = ET.parse(file_path)
        root = tree.getroot()
        data = []
        for element in root:
            element_data = {}
            for child in element:
                element_data[child.tag] = child.text
            data.append(element_data)
        return data

# Factory class for creating data readers
class DataReaderFactory:
    @staticmethod
    def create_data_reader(file_path):
        file_extension = os.path.splitext(file_path)[1].lower()
        if file_extension == ".csv":
            return CSVDataReader()
        elif file_extension == ".json":
            return JSONDataReader()
        elif file_extension == ".xml":
            return XMLDataReader()
        else:
            raise ValueError("Unsupported file type")

# Client code
if __name__ == "__main__":
    if len(sys.argv) != 2:
        print("Usage: python script.py <file_path>")
        sys.exit(1)

    file_path = sys.argv[1]
    
    data_reader = DataReaderFactory.create_data_reader(file_path)
    data = data_reader.read_data(file_path)
    print("Data read from", file_path, ":", data)